src/Servers/HttpSys/src/RequestProcessing/Response.cs

```diff
@@ -527,128 +516,92 @@ private static bool Matches(string knownValue, StringValues input)
-                        pinnedHeaders.Add(gcHandle);
-                        header.KnownHeaders = (HttpApiTypes.HTTP_KNOWN_HEADER*)gcHandle.AddrOfPinnedObject();
-
-                        for (int headerValueIndex = 0; headerValueIndex < headerValues.Count; headerValueIndex++)
-                        {
-                            // Add Value
-                            headerValue = headerValues[headerValueIndex] ?? string.Empty;
-                            bytes = HeaderEncoding.GetBytes(headerValue);
-                            nativeHeaderValues[header.KnownHeaderCount].RawValueLength = (ushort)bytes.Length;
-                            gcHandle = GCHandle.Alloc(bytes, GCHandleType.Pinned);
-                            pinnedHeaders.Add(gcHandle);
-                            nativeHeaderValues[header.KnownHeaderCount].pRawValue = (byte*)gcHandle.AddrOfPinnedObject();
-                            header.KnownHeaderCount++;
-                        }
-
-                        // This type is a struct, not an object, so pinning it causes a boxed copy to be created. We can't do that until after all the fields are set.
-                        gcHandle = GCHandle.Alloc(header, GCHandleType.Pinned);
-                        pinnedHeaders.Add(gcHandle);
-                        knownHeaderInfo[_nativeResponse.ResponseInfoCount].pInfo = (HttpApiTypes.HTTP_MULTIPLE_KNOWN_HEADERS*)gcHandle.AddrOfPinnedObject();
-
-                        _nativeResponse.ResponseInfoCount++;
+                        // Add Value
+                        headerValue = headerValues[headerValueIndex] ?? string.Empty;
+                        bytes = allocator.GetHeaderEncodedBytes(headerValue, out bytesLength);
+                        nativeHeaderValues[header->KnownHeaderCount].RawValueLength = checked((ushort)bytesLength);
+                        nativeHeaderValues[header->KnownHeaderCount].pRawValue = bytes;
+                        header->KnownHeaderCount++;
                     }
-                }
-            }
-        }
-        catch
-        {
-            FreePinnedHeaders(pinnedHeaders);
-            throw;
-        }
-        return pinnedHeaders;
-    }
 
-    private static void FreePinnedHeaders(List<GCHandle>? pinnedHeaders)
-    {
-        if (pinnedHeaders != null)
-        {
-            foreach (GCHandle gcHandle in pinnedHeaders)
-            {
-                if (gcHandle.IsAllocated)
-                {
-                    gcHandle.Free();
+                    knownHeaderInfo[_nativeResponse.ResponseInfoCount].pInfo = header;
+
+                    _nativeResponse.ResponseInfoCount++;
                 }
             }
         }
     }
 
-    internal unsafe void SerializeTrailers(HttpApiTypes.HTTP_DATA_CHUNK[] dataChunks, int currentChunk, List<GCHandle> pins)
+    internal unsafe void SerializeTrailers(ref UnmanagedBufferAllocator allocator, Span<HttpApiTypes.HTTP_DATA_CHUNK> dataChunks, int currentChunk)
```

Comment:
I avoided this logic because of the assert within `SerializeTrailers`. My reading of this was the "trailers" must be at the end so the assert of that case should occur within that function as opposed to it's callers. Since there is only one caller I guess this is okay, but I'm not convinced this is worth the change as it spreads a seemingly important invariant to all potential call sites.

Reply:
I have been reviewing the code to make sure that the exact right number of chunks is initialized. 

It made me to suggest the assert at the end https://github.com/dotnet/aspnetcore/pull/44561#discussion_r997732383 and also this change. As written, it is not clear whether `SerializeTrailers` fills in one dataChunk or multiple dataChunks.